src/server/pfs/server/source.go

```diff
@@ -0,0 +1,255 @@
+		}
+		fr.fileCount--
+	}
+	if len(noPadding) > 0 && noPadding[0] {
+		return nil
+	}
+	// Close a tar writer to create tar EOF padding.
+	return tar.NewWriter(w).Close()
+}
+
+func (fr *FileReader) drain() error {
+	for fr.fileCount > 0 {
+		if _, err := fr.mr.Next(); err != nil {
+			return err
+		}
+		fr.fileCount--
+	}
+	return nil
+}
+
+// Source iterates over FileInfoV2s generated from a fileset.Source
+type Source struct {
+	commit        *pfs.Commit
+	getReader     func() fileset.FileSource
+	computeHashes bool
+}
+
+// Creates a Reader which emits FileInfoV2s with the information from commit, and the entries from readers
+// returned by getReader.  If getReader returns different Readers all bets are off.
+func NewSource(commit *pfs.Commit, computeHashes bool, getReader func() fileset.FileSource) *Source {
+	return &Source{
+		commit:        commit,
+		getReader:     getReader,
+		computeHashes: computeHashes,
+	}
+}
+
+func (s *Source) Iterate(ctx context.Context, cb func(*pfs.FileInfoV2, fileset.File) error) error {
+	ctx, cf := context.WithCancel(ctx)
+	defer cf()
+	fs1 := s.getReader()
+	fs2 := s.getReader()
+	s2 := newStream(ctx, fs2)
+	cache := make(map[string][]byte)
+	return fs1.Iterate(ctx, func(fr fileset.File) error {
+		idx := fr.Index()
+		fi := &pfs.FileInfoV2{
+			File: client.NewFile(s.commit.Repo.Name, s.commit.ID, idx.Path),
+		}
+		if s.computeHashes {
+			var err error
+			var hashBytes []byte
+			if indexIsDir(idx) {
+				hashBytes, err = computeHash(cache, s2, idx.Path)
+				if err != nil {
+					return err
+				}
+			} else {
+				hashBytes = computeFileHash(idx)
+			}
+			fi.Hash = pfs.EncodeHash(hashBytes)
+		}
+		if err := cb(fi, fr); err != nil {
+			return err
+		}
+		delete(cache, idx.Path)
+		return nil
+	})
+}
+
+func computeHash(cache map[string][]byte, s *stream, x string) ([]byte, error) {
+	if hashBytes, exists := cache[x]; exists {
+		return hashBytes, nil
+	}
+	// check that the next item in the stream is x
+	fr, err := s.Peek()
+	if err != nil {
+		if err == io.EOF {
+			return nil, errors.Errorf("stream is done, can't compute hash for %s", x)
+		}
+		return nil, err
+	}
+	idx := fr.Index()
+	if idx.Path != x {
+		return nil, errors.Errorf("stream is wrong place to compute hash for %s", x)
+	}
+	// consume x from the stream
+	fr, err = s.Next()
+	if err != nil {
+		return nil, err
+	}
+	idx = fr.Index()
+	// for file
+	if !indexIsDir(idx) {
+		return computeFileHash(idx), nil
+	}
+	// for directory
+	h := pfs.NewHash()
+	for {
+		f2, err := s.Peek()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		idx2 := f2.Index()
+		if !strings.HasPrefix(idx2.Path, x) {
+			break
+		}
+		childHash, err := computeHash(cache, s, idx2.Path)
+		if err != nil {
+			return nil, err
+		}
+		h.Write(childHash)
+	}
+	hashBytes := h.Sum(nil)
+	cache[x] = hashBytes
+	return hashBytes, nil
+}
+
+func computeFileHash(idx *index.Index) []byte {
+	h := pfs.NewHash()
+	if idx.DataOp != nil {
```

Comment:
Lets go ahead and get the fix in for full chunk references. `dataRef.Hash` will be `""` if it is a full chunk reference, so the hash should be pulled from `dataRef.ChunkInfo.Chunk.Hash` in that case.